Windows: make `Command` prefer non-verbatim paths

ChrisDenton opened the pull request:
When spawning Commands, the path we use can end up being queried using `env::current_exe` (or the equivalent in other languages). Not all applications handle these paths properly therefore we should have a stronger preference for non-verbatim paths when spawning processes.

rust-highfive commented:
Hey! It looks like you've submitted a new PR for the library teams!

If this PR contains changes to any `rust-lang/rust` public library APIs then please comment with `r? rust-lang/libs-api @rustbot label +T-libs-api -T-libs` to request review from a libs-api team reviewer. If you're unsure where your change falls no worries, just leave it as is and the reviewer will take a look and make a decision to forward on if necessary.

Examples of `T-libs-api` changes:

* Stabilizing library features
* Introducing insta-stable changes such as new implementations of existing stable traits on existing stable types
* Introducing new or changing existing unstable library APIs (excluding permanently unstable features / features without a tracking issue)
* Changing public documentation in ways that create new stability guarantees
* Changing observable runtime behavior of library APIs

ChrisDenton commented:
Oh, this does have observable changes.

r? rust-lang/libs-api @rustbot label +T-libs-api -T-libs

joshtriplett commented:
@rfcbot merge

bors commented:
:umbrella: The latest upstream changes (presumably #104160) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

matklad commented:
Just in case, cc @kornelski, the author of https://lib.rs/crates/dunce

kornelski commented:
DOS paths have not only character length limit, but also don't support reserved names like `com` and don't allow special characters like `:`.

Feel free to copy dunce if you need: https://gitlab.com/kornelski/dunce/-/blob/master/src/lib.rs#L163

ChrisDenton committed:
Windows: make Command prefer non-verbatim paths

When spawning Commands, the path we use can end up being queried using `env::current_exe` (or the equivalent in other languages). Not all applications handle these paths properly therefore we should have a stronger preference for non-verbatim paths when spawning processes.

ChrisDenton commented:
Note that the code being called here is doing the opposite of dunce, it's adding `\\?\`. The change in this PR is to make it slightly less aggressive about adding `\\?\` for paths that may end up being visible to other user code.

ChrisDenton commented:
Ah, this passed fcp months back but was never merged. Sorry about that; the holiday season and into the new year were busy for me. 

@bors r=joshtriplett

workingjubilee commented:
@bors r=joshtriplett

bors commented:
:pushpin: Commit 920435f195b49029d0a83679c04d7dda66afb50a has been approved by `joshtriplett`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joshtriplett 920435f195b49029d0a83679c04d7dda66afb50a -->
<!-- homu: {"type":"Approved","sha":"920435f195b49029d0a83679c04d7dda66afb50a","approver":"joshtriplett","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
